### ISSUE TITLE
Authenticate API access for more fine-grained rate-limiting

### ISSUE THREAD

**Issue by paulbellamy** [stellar/go#4311]
### What problem does your feature solve?

Differentiating and rate-limiting is very coarse when users are all anonymous.

### What would you like to see?

Add an authentication mechanism to the API, so we can do more fine-grained rate-limiting.

We would need to:
1. Make an announcement
1. Add a way to authenticate
1. Update SDKs to do that
1. Add harsher rate-limits for unauthenticated users, higher for authenticated/preferred/etc.

This could (eventually) open the door towards other people running (and charging for) their own horizon instances.

### What alternatives are there?

previous discussion: https://github.com/stellar/go/issues/4263#issuecomment-1079232009

Api gateway lets you plug in an authenticator, so we could sign a jwt with the users’ stellar private key, and authenticate on that. (as a suggestion). Re-using stellar accounts as an authentication mechanism is an appealing mechanism.

**Comment by sreuland** [comment]
related to authenticator lower level mechanics, can leverage JWT standard for authentication with signatures based on Stellar Accounts for identity.
 
* client creates their own Stellar Account, which establishes a key-pair by default. 
* client creates and signs JWT, placing their stellar public key(encoded as StrKey) as an identity claim in the payload and signing JWT with their stellar account private key and puts the base64 of JWT into HTTP `Authorization` header
* An HTTP intermediate server that supports pluggable authenticators(proxy or gateway) or Horizon itself can execute the authenticator to verify the JWT on incoming requests.
* JWT verification:
  * service checks JWT payload for identity claim like standard key`sub` or can be custom key like `stellar_account` 
  * service verifies JWT signature with the identity claim(the stellar public key) or rejects if bad
  * service queries an internal whitelist datastore for a match on the identity claim(stellar public key encoded as StrKey) or rejects if not-found.
    * will likely require some persistence design for where the 'whitelist' of allowed stellar accounts is stored. 
    * could start to consider more advanced rules or ACLs per account, 'requests-per-minute', etc. ..

**Comment by leighmcculloch** [comment]
> * client creates their own Stellar Account, which establishes a key-pair by default.
> * client creates and signs JWT

This is elegant. Stellar keys are ed25519 (eddsa), and eddsa isn't widely supported in jwt libs, at least according to jwt.io, so compatibility might be limited. Would Horizon do a check that the account exists on every request? If yes, that'd require a db lookup. The main advantage with JWTs is that you can check the signature inline of a request with zero external requests. So ideally the user would auth to another server first and receive a reusable JWT signed by a key owned by the Horizon operator.

Much of what is described sounds similar to [SEP-10], which uses Stellar accounts. 

We have a SEP-10 implementation in this repo already, that we're using in production already, and from Horizons perspective it would still just be a JWT to verify. https://github.com/stellar/go/tree/master/exp/services/webauth

**Comment by sreuland** [comment]
> eddsa isn't widely supported in jwt libs, at least according to jwt.io, so compatibility might be limited

good point, would need to take inventory on jwt lib support, briefly looking, some libs/languages showing how to use or extend lib to allow using eddsa.

> Would Horizon do a check that the account exists on every request?

The identity `sub` claim from JWT exists as a stellar account on ledger? That could be optional config to enforce during `authorizor` server authentication, which may take place in Horizon or possibly before it.

initially thinking about JWT for Horizon API access sent by caller as lightweight, `bearer` provided, so caller creates in one step with `sub`(identity)=`iss`(issuer), rather than having flow with intermediary issuer of JWTs. The `sub` represents an entity( service/app/org/individual/etc) rather than an app user, and the notion of expiration `iat` isn't super important, the JWT could have longer lifetime, so caller doesn't have to rebuild it regularly. Maybe using the SEP10 auth flow with callers going through challenge flow to get shorter lived JWT repeatedly is another way to get it done.

lifecycle highlights:
* one time: an entity creates a stellar account to represent itself as a service/client.
* one time: enrollment, likely out-of-band, entity requests and is accepted by host of Horizon API to be allowed as a client, maybe SEP10 challenge flow can automate this exchange?
* entity's apps/services/scripts generate JWT with edd25519 key pair from their stellar account and include as `Authorization` header when sending HTTP requests to Horizon API.
* `authorizor` a service or module invoked by LB/Proxy/API Gateway or inside Horizon(go middleware). Receives HTTP request first and does the `inline` JWT sig validation, passing validated JWT to `whitelist` service
* `whitelist` could be service or module invoked by LB/Proxy/API Gateway or inside Horizon. Queries some form of persistence  that stores 'allowed' clients as strkey account id. Persistence format could be db, file(JWKS file with JWK for each allowed client key), url, needs to be maintained by Horizon host.

**Comment by sreuland** [comment]
@jcx120 , this is the dev ticket tracking API authentication design.
